File: algorithm/baekjoon/src/main/java/com/keunwon/algorithm/dfs/Problem1260.java
package com.keunwon.algorithm.dfs;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Problem1260 {

    private static int V;
    private static int E;
    private static int S;
    private static int[][] nodes;
    private static boolean[] checked;

    public static void main(String[] args) throws IOException {
        input();

        dfs(S);
        resetChecked();
        bfs();
    }

    private static void dfs(int x) {
        System.out.print(x + " ");
        
        checked[x] = true;

        for (int i = 1; i < V + 1; i++) {
            if (nodes[x][i] == 1 && !checked[i]) {
                dfs(i);
            }
        }
    }

    private static void bfs() {
        Queue<Integer> queue = new LinkedList<>();
        queue.add(S);
        checked[S] = true;

        System.out.print(System.lineSeparator() + S + " ");

        while (!queue.isEmpty()) {
            int temp = queue.poll();

            for (int i = 0; i < V + 1; i++) {
                if (nodes[temp][i] == 1 && !checked[i]) {
                    queue.add(i);
                    checked[i] = true;
                    System.out.print(i + " ");
                }
            }
        }
    }

    private static void resetChecked() {
        checked = new boolean[V + 1];
    }

    private static void input() throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        V = Integer.parseInt(st.nextToken());
        E = Integer.parseInt(st.nextToken());
        S = Integer.parseInt(st.nextToken());
        nodes = new int[V + 1][V + 1];
        checked = new boolean[V + 1];

        for (int i = 0; i < E; i++) {
            st = new StringTokenizer(br.readLine());
            int x = Integer.parseInt(st.nextToken());
            int y = Integer.parseInt(st.nextToken());

            for (int j = 1; j < V + 1; j++) {
                if (j == y) {
                    nodes[x][j] = nodes[j][x] = 1;
                }
            }
        }
    }
}

File: algorithm/kotlin-algorithm/baekjoon/src/main/kotlin/algorithm/baekjoon/Problem1260.kt
package algorithm.baekjoon

import java.util.LinkedList

/**
 * <p>
 * 이름: DFS와 BFS
 * 난이도: 실버-2
 * </p>
 *
 * @see <a href="https://www.acmicpc.net/problem/1260">DFS와 BFS (실버-2)</a>
 **/
class Problem1260 {
    private lateinit var grpah: Array<BooleanArray>
    private val result = List(2) { mutableListOf<Int>() }

    fun solution(n: Int, v: Int, edges: Array<Pair<Int, Int>>): List<List<Int>> {
        this.grpah = Array(n + 1) { BooleanArray(n + 1) }.apply {
            edges.forEach { (a, b) ->
                this[a][b] = true
                this[b][a] = true
            }
        }

        dfs(v, BooleanArray(n + 1))
        bfs(v, n)

        return result
    }

    private fun bfs(v: Int, n: Int) {
        val q = LinkedList<Int>().apply { offer(v) }
        val visited = BooleanArray(n + 1).apply { this[v] = true }

        while (q.isNotEmpty()) {
            val cur = q.poll()
            result[1].add(cur)

            for (i in 1..n) {
                if (grpah[cur][i] && !visited[i]) {
                    visited[i] = true
                    q.offer(i)
                }
            }
        }
    }

    private fun dfs(v: Int, visited: BooleanArray) {
        visited[v] = true
        result[0].add(v)

        for (i in 1 until visited.size) {
            if (grpah[v][i] && !visited[i]) dfs(i, visited)
        }
    }
}

fun main() {
    val br = System.`in`.bufferedReader()
    val (n, m, v) = br.readLine().split(" ").map { it.toInt() }
    val edges = Array(m) {
        val arr = br.readLine().split(" ").map { it.toInt() }
        arr[0] to arr[1]
    }

    Problem1260().solution(n, v, edges).forEach { println(it.joinToString(" ")) }
}
